File: backend/Services/InterviewService.cs
using System.Text.RegularExpressions;
using CareerMail.Api.Data;
using CareerMail.Api.Models.DTOs;
using CareerMail.Api.Models.Entities;
using CareerMail.Api.Models.Enums;
using Microsoft.EntityFrameworkCore;

namespace CareerMail.Api.Services;

public interface IInterviewService
{
    Task<List<Interview>> GetAllInterviewsAsync(long userId);
    Task<Interview> GetInterviewByIdAsync(long userId, long id);
    Task<Interview> CreateInterviewAsync(long userId, InterviewRequest request);
    Task<Interview> UpdateInterviewAsync(long userId, long id, InterviewRequest request);
    Task DeleteInterviewAsync(long userId, long id);
}

public class InterviewService : IInterviewService
{
    private readonly AppDbContext _context;

    public InterviewService(AppDbContext context)
    {
        _context = context;
    }

    public async Task<List<Interview>> GetAllInterviewsAsync(long userId)
    {
        var list = await _context.Interviews
            .Include(i => i.JobApplication)
            .Where(i => i.UserId == userId)
            .OrderBy(i => i.InterviewDate)
            .ToListAsync();

        list.ForEach(ComputeDaysAwayBadge);
        return list;
    }

    public async Task<Interview> GetInterviewByIdAsync(long userId, long id)
    {
        var interview = await _context.Interviews
            .Include(i => i.JobApplication)
            .FirstOrDefaultAsync(i => i.Id == id && i.UserId == userId);

        if (interview == null)
        {
            throw new ArgumentException($"Interview not found with ID: {id}");
        }

        ComputeDaysAwayBadge(interview);
        return interview;
    }

    public async Task<Interview> CreateInterviewAsync(long userId, InterviewRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.Company)) throw new ArgumentException("Company is required");
        if (string.IsNullOrWhiteSpace(request.Title)) throw new ArgumentException("Title is required");
        if (!request.InterviewDate.HasValue) throw new ArgumentException("Interview date is required");

        var interview = new Interview
        {
            UserId = userId,
            JobApplicationId = request.JobApplicationId,
            Company = request.Company.Trim(),
            Title = request.Title.Trim(),
            InterviewDate = request.InterviewDate.Value,
            Type = !string.IsNullOrWhiteSpace(request.Type) ? request.Type.Trim() : "Technical Interview",
            Interviewer = request.Interviewer?.Trim(),
            Location = !string.IsNullOrWhiteSpace(request.Location) ? request.Location.Trim() : "Google Meet",
            MeetingLink = request.MeetingLink?.Trim(),
            PreparationNotes = request.PreparationNotes?.Trim(),
            Status = InterviewStatusExtensions.FromString(request.Status),
            CompanyLogo = !string.IsNullOrWhiteSpace(request.CompanyLogo)
                ? request.CompanyLogo.Trim()
                : Regex.Replace(request.Company.ToLowerInvariant(), "[^a-z0-9]", "")
        };

        if (request.JobApplicationId.HasValue)
        {
            var app = await _context.JobApplications.FirstOrDefaultAsync(ja => ja.Id == request.JobApplicationId.Value && ja.UserId == userId);
            if (app != null)
            {
                app.TimelineEvents.Add(new TimelineEvent
                {
                    Title = $"Interview Scheduled: {interview.Type}",
                    Description = $"Date: {interview.InterviewDate} with {interview.Interviewer ?? "Hiring Team"}",
                    EventDate = DateTime.UtcNow,
                    EventType = "INTERVIEW"
                });
            }
        }

        ComputeDaysAwayBadge(interview);
        _context.Interviews.Add(interview);
        await _context.SaveChangesAsync();

        return interview;
    }

    public async Task<Interview> UpdateInterviewAsync(long userId, long id, InterviewRequest request)
    {
        var interview = await GetInterviewByIdAsync(userId, id);

        if (!string.IsNullOrWhiteSpace(request.Company)) interview.Company = request.Company.Trim();
        if (!string.IsNullOrWhiteSpace(request.Title)) interview.Title = request.Title.Trim();
        if (request.InterviewDate.HasValue) interview.InterviewDate = request.InterviewDate.Value;
        if (!string.IsNullOrWhiteSpace(request.Type)) interview.Type = request.Type.Trim();
        if (request.Interviewer != null) interview.Interviewer = request.Interviewer.Trim();
        if (request.Location != null) interview.Location = request.Location.Trim();
        if (request.MeetingLink != null) interview.MeetingLink = request.MeetingLink.Trim();
        if (request.PreparationNotes != null) interview.PreparationNotes = request.PreparationNotes.Trim();
        if (!string.IsNullOrWhiteSpace(request.Status)) interview.Status = InterviewStatusExtensions.FromString(request.Status);
        if (!string.IsNullOrWhiteSpace(request.CompanyLogo)) interview.CompanyLogo = request.CompanyLogo.Trim();
        if (request.JobApplicationId.HasValue) interview.JobApplicationId = request.JobApplicationId.Value;

        ComputeDaysAwayBadge(interview);
        await _context.SaveChangesAsync();
        return interview;
    }

    public async Task DeleteInterviewAsync(long userId, long id)
    {
        var interview = await GetInterviewByIdAsync(userId, id);
        _context.Interviews.Remove(interview);
        await _context.SaveChangesAsync();
    }

    private static void ComputeDaysAwayBadge(Interview interview)
    {
        var now = DateTime.UtcNow;
        if (interview.InterviewDate < now)
        {
            interview.DaysAwayBadge = "Completed";
            return;
        }

        var diff = interview.InterviewDate - now;
        if (diff.TotalDays < 1)
        {
            var hours = (int)diff.TotalHours;
            interview.DaysAwayBadge = hours <= 0 ? "Today" : $"In {hours}h";
        }
        else if (diff.TotalDays < 2)
        {
            interview.DaysAwayBadge = "Tomorrow";
        }
        else
        {
            interview.DaysAwayBadge = $"In {(int)diff.TotalDays} days";
        }
    }
}

File: backend/app/services/interview_service.py
import re
from datetime import datetime
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from app.models.models import Interview, JobApplication, TimelineEvent
from app.models.enums import InterviewStatus
from app.schemas.interview import InterviewRequest


class InterviewService:
    def __init__(self, db: Session):
        self.db = db

    def get_all_interviews(self, user_id: int) -> list[Interview]:
        interviews = (
            self.db.query(Interview)
            .filter(Interview.user_id == user_id)
            .order_by(Interview.interview_date.asc())
            .all()
        )
        for i in interviews:
            self._compute_days_away_badge(i)
        return interviews

    def get_interview_by_id(self, user_id: int, interview_id: int) -> Interview:
        interview = (
            self.db.query(Interview)
            .filter(Interview.id == interview_id, Interview.user_id == user_id)
            .first()
        )
        if not interview:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Interview not found with ID: {interview_id}"
            )
        self._compute_days_away_badge(interview)
        return interview

    def create_interview(self, user_id: int, request: InterviewRequest) -> Interview:
        if not request.company or not request.company.strip():
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Company is required")
        if not request.title or not request.title.strip():
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Title is required")
        if not request.interview_date:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Interview date is required")

        company = request.company.strip()
        logo = request.company_logo.strip() if request.company_logo and request.company_logo.strip() else re.sub(r"[^a-z0-9]", "", company.lower())

        interview = Interview(
            user_id=user_id,
            job_application_id=request.job_application_id,
            company=company,
            title=request.title.strip(),
            interview_date=request.interview_date,
            type=request.type.strip() if request.type else "Technical Interview",
            interviewer=request.interviewer.strip() if request.interviewer else None,
            location=request.location.strip() if request.location else "Google Meet",
            meeting_link=request.meeting_link.strip() if request.meeting_link else None,
            preparation_notes=request.preparation_notes.strip() if request.preparation_notes else None,
            status=InterviewStatus.from_string(request.status).value,
            company_logo=logo
        )

        if request.job_application_id:
            app = (
                self.db.query(JobApplication)
                .filter(JobApplication.id == request.job_application_id, JobApplication.user_id == user_id)
                .first()
            )
            if app:
                app.timeline_events.append(TimelineEvent(
                    title=f"Interview Scheduled: {interview.type}",
                    description=f"Date: {interview.interview_date} with {interview.interviewer or 'Hiring Team'}",
                    event_date=datetime.utcnow(),
                    event_type="INTERVIEW"
                ))

        self._compute_days_away_badge(interview)
        self.db.add(interview)
        self.db.commit()
        self.db.refresh(interview)
        return interview

    def update_interview(self, user_id: int, interview_id: int, request: InterviewRequest) -> Interview:
        interview = self.get_interview_by_id(user_id, interview_id)

        if request.company and request.company.strip():
            interview.company = request.company.strip()
        if request.title and request.title.strip():
            interview.title = request.title.strip()
        if request.interview_date is not None:
            interview.interview_date = request.interview_date
        if request.type is not None:
            interview.type = request.type.strip()
        if request.interviewer is not None:
            interview.interviewer = request.interviewer.strip() if request.interviewer else None
        if request.location is not None:
            interview.location = request.location.strip()
        if request.meeting_link is not None:
            interview.meeting_link = request.meeting_link.strip() if request.meeting_link else None
        if request.preparation_notes is not None:
            interview.preparation_notes = request.preparation_notes.strip() if request.preparation_notes else None
        if request.status and request.status.strip():
            interview.status = InterviewStatus.from_string(request.status).value
        if request.company_logo is not None:
            interview.company_logo = request.company_logo.strip()
        if request.job_application_id is not None:
            interview.job_application_id = request.job_application_id

        self._compute_days_away_badge(interview)
        self.db.commit()
        self.db.refresh(interview)
        return interview

    def delete_interview(self, user_id: int, interview_id: int):
        interview = self.get_interview_by_id(user_id, interview_id)
        self.db.delete(interview)
        self.db.commit()

    @staticmethod
    def _compute_days_away_badge(interview: Interview):
        now = datetime.utcnow()
        if interview.interview_date < now:
            interview.days_away_badge = "Completed"
            return

        diff = interview.interview_date - now
        if diff.total_seconds() < 86400:
            hours = int(diff.total_seconds() // 3600)
            interview.days_away_badge = "Today" if hours <= 0 else f"In {hours}h"
        elif diff.total_seconds() < 172800:
            interview.days_away_badge = "Tomorrow"
        else:
            days = int(diff.total_seconds() // 86400)
            interview.days_away_badge = f"In {days} days"
